arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/EagerEffect.kt

```diff
@@ -0,0 +1,114 @@
+package arrow.core.continuations
+
+import arrow.core.Either
+import arrow.core.Ior
+import arrow.core.Option
+import arrow.core.Some
+import arrow.core.Validated
+import arrow.core.identity
+import arrow.core.nonFatalOrThrow
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.EmptyCoroutineContext
+import kotlin.coroutines.intrinsics.startCoroutineUninterceptedOrReturn
+
+public interface EagerEffect<R, A> {
+  public fun <B> fold(recover: (R) -> B, transform: (A) -> B): B
+
+  public fun <B> fold(
+    error: (error: Throwable) -> B,
+    recover: (shifted: R) -> B,
+    transform: (value: A) -> B
+  ): B =
+    try {
+      fold(recover, transform)
+    } catch (e: Throwable) {
+      error(e.nonFatalOrThrow())
+    }
+
+  public fun toIor(): Ior<R, A> = fold({ Ior.Left(it) }) { Ior.Right(it) }
+
+  public fun toEither(): Either<R, A> = fold({ Either.Left(it) }) { Either.Right(it) }
+
+  public fun toValidated(): Validated<R, A> =
+    fold({ Validated.Invalid(it) }) { Validated.Valid(it) }
+
+  public fun orNull(): A? = fold({ null }, ::identity)
+
+  public fun toOption(orElse: (R) -> Option<A>): Option<A> =
+    fold(orElse, ::Some)
+
+  public fun <B> map(f: (A) -> B): EagerEffect<R, B> = flatMap { a -> eagerEffect { f(a) } }
+
+  public fun <B> flatMap(f: (A) -> EagerEffect<R, B>): EagerEffect<R, B> = eagerEffect {
+    f(bind()).bind()
+  }
+
+  public fun attempt(): EagerEffect<R, Result<A>> = eagerEffect {
+    kotlin.runCatching { bind() }
+  }
+
+  public fun handleError(f: (R) -> A): EagerEffect<Nothing, A> = eagerEffect {
+    fold(f, ::identity)
+  }
+
+  public fun <R2> handleErrorWith(f: (R) -> EagerEffect<R2, A>): EagerEffect<R2, A> =
+    eagerEffect {
+      toEither().fold({ r -> f(r).bind() }, ::identity)
+    }
+
+  public fun <B> redeem(f: (R) -> B, g: (A) -> B): EagerEffect<Nothing, B> = eagerEffect {
+    fold(f, g)
+  }
+
+  public fun <R2, B> redeemWith(
+    f: (R) -> EagerEffect<R2, B>,
+    g: (A) -> EagerEffect<R2, B>
+  ): EagerEffect<R2, B> = eagerEffect { fold(f, g).bind() }
+}
+
+@PublishedApi
+internal class Eager(val token: Token, val shifted: Any?, val recover: (Any?) -> Any?) :
+  ShiftCancellationException() {
+  override fun toString(): String = "ShiftCancellationException($message)"
+}
```

Comment:
I think we should use the same trick here as we do for `Effect`, so users have an escape hatch to differentiate between `JobCanncellationException` and `ShiftCanncelationException` when they have to in low-level niche use-cases.

Reply:
added tests in EagerEffect and Effect for low-level use-cases